### PR TITLE
Add select functionality

### PR DESCRIPTION
**Pull request by samelamin**
I tried doing this test first but failed miserably because I wanted it to be an extention on the sqlContext

Any ideas how best to test it?

### CODE REVIEW

**Comment by holdenk** [comment]
So if you add the `package com.appsflyer.spark.bigquery` to the top of `BigQuerySchemaSpecs` then it will have the implicits in scope (or you can import the implicits explicitily if you need access to them in another scope) and then it should just work.

**Comment by samelamin** [comment]
@holdenk thanks for helping out here, really appreciate it! I was thinking of maybe testing the BigQueryClient instead, that way we can inject the bqService and mock it out? Do you think that might be a cleaner approach?

**Comment by holdenk** [comment]
I'd probably refactor your code a bit to make it easier to supply an explicit mocked `BigQueryClient`, but that's a matter of personal taste.

**Comment by samelamin** [comment]
Actually just thinking about it, if we do it your way we will be mocking the BigQuery client but then it raises the question of what exactly are we testing, by mocking out the bigqueryclient we will be just testing sparks ability to read Json, While the majority of the logic in the BigQuery client will be untested

Wouldnt it be better to to mock out the bigquery service (dependency of the BigQueryClient) and hence have a black box test that exercises all parts of the system? 

Not arguing here, just trying to understand. Its not every day I get a chance to learn from an expert ;)

**Comment by holdenk** [comment]
Thanks, that's very nice to say, although my expertise is a bit narrow with regards to this so feel free to ignore any of my advice which seems off-base :) So it all comes down to which part you want to test, if you want to verify your calls from the bigquery client to the service then you probably still want to refactor your code to make it easier to specify a custom bigquery client, except instead of providing a mock bigquery client you will provide a custom bigquery client with a mock bigquery service object.

You can of course create a testing only sub-class with the stub but its up to you.

**Comment by samelamin** [comment]
@holdenk how does it look now?


@spektom any thoughts?

**Comment by spektom** [comment]
@samelamin Looks fine!

File: src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala [hunk 1]
```diff
@@ -9,56 +9,49 @@ import com.google.api.client.http.javanet.NetHttpTransport
...
   private val logger = LoggerFactory.getLogger(classOf[BigQueryClient])
-
   private val SCOPES = List(BigqueryScopes.BIGQUERY).asJava
 
-  private val bigquery: Bigquery = {
-    val credential = GoogleCredential.getApplicationDefault.createScoped(SCOPES)
-    new Bigquery.Builder(new NetHttpTransport, new JacksonFactory, credential)
-      .setApplicationName("spark-bigquery")
-      .build()
+  if(bigquery == null) {
```
Comment (holdenk) [inline src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala]: Do we expect this to need to support multi-threaded use cases? If so we might have a race condition here.

File: src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala [hunk 2]
```diff
@@ -9,56 +9,49 @@ import com.google.api.client.http.javanet.NetHttpTransport
...
+import org.slf4j.LoggerFactory
+
 import scala.util.Random
 import scala.util.control.NonFatal
 
 /**
   * Created by sam elamin on 11/01/2017.
   */
-private[bigquery] object BigQueryClient {
+class BigQueryClient(sqlContext: SQLContext, var bigquery: Bigquery = null) {
```
Comment (holdenk) [inline src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala]: Minor point, but I'd mention what the expected use case of this is (e.g. do you want people to call this directly? Is there any reason besides testing someone might want to specify the second param? etc.)

File: src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala [hunk 3]
```diff
@@ -89,7 +82,20 @@ private[bigquery] class BigQueryClient(conf: Configuration) {
...
+    val fullyQualifiedInputTableId = BigQueryStrings.toString(tableReference)
+    BigQueryConfiguration.configureBigQueryInput(hadoopConf, fullyQualifiedInputTableId)
+
+    val tableData = sqlContext.sparkContext.newAPIHadoopRDD(
+      hadoopConf,
+      classOf[GsonBigQueryInputFormat],
+      classOf[LongWritable],
+      classOf[JsonObject]).map(_._2.toString)
+
+    val df = sqlContext.read.json(tableData)
```
Comment (holdenk) [inline src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala]: This is a semi-expensive operation - I don't know much about the big query interface but is there a way to get some schema information from this and speed it up? Might not be needed for this PR but if there is maybe leave this as a TODO comment for someone to get to later (or create a follow up GH/JIRA issue)?

File: src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala [hunk 4]
```diff
@@ -108,7 +114,7 @@ private[bigquery] class BigQueryClient(conf: Configuration) {
       case e: GoogleJsonResponseException if e.getStatusCode == 404 =>
         logger.info(s"Creating staging dataset $projectId:$datasetId")
         val dsRef = new DatasetReference().setProjectId(projectId).setDatasetId(datasetId)
-        val ds = new Dataset()
+        val ds = new  com.google.api.services.bigquery.model.Dataset()
```
Comment (holdenk) [inline src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala]: +1 for disambiguiting which dataset (could also use import renaming but matter of personal choice).

File: src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala (hunk 2 above)
Comment (samelamin) [inline src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala]: Tbh i only added to enable me to test, but cant think of a way to enable injecting other than this

in most cases people wont be adding it

Any ideas how else I can do it?

File: src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala (hunk 3 above)
Comment (samelamin) [inline src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala]: We might be able to get the schema information from the JSONObject, but how will that speed it up? create the dataframe using the schema? 

How is it an expensive operation?

File: src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala (hunk 4 above)
Comment (samelamin) [inline src/main/scala/com/appsflyer/spark/bigquery/BigQueryClient.scala]: I agree, i prefer import renaming. ill change it

**Comment by samelamin** [comment]
@holdenk no worries on the delay, I really appreciate whatever time you put into this and if I can ever repay the favour please do not hesitate to get in touch

I merged the PR in knowing full well that a kitten dies everytime one merges their on pr 😢

**Pull request merged**